url73e3057d61_empeeoh_BACnet4J_tgz-pJ8-com_serotonin_bacnet4j_test_OverflowTestJ8/src/com/serotonin/bacnet4j/util/RequestUtils.java

```diff
@@ -0,0 +1,400 @@
+                            updater.increment(oid, result.getPropertyIdentifier(), 
+                            				  result.getPropertyArrayIndex(),
+                            				  result.getReadResult().getDatum());
+                            if (updater.cancelled())
+                                break;
+                        }
+
+                        if (updater.cancelled())
+                            break;
+                    }
+                }
+                catch (AbortAPDUException e) {
+                    LOG.warn("Chunked request failed.");
+                    if (e.getApdu().getAbortReason() == AbortReason.bufferOverflow.intValue()
+                     || e.getApdu().getAbortReason() == AbortReason.segmentationNotSupported.intValue()) {
+                        if (counter > 0)
+                            sendOneAtATime(localDevice, rd, partition, updater);
+                        else {
+                            // Failed on the first partition. Send all one at a time, reduce the device's max
+                            // references, and quit.
+                            sendOneAtATime(localDevice, rd, refs, updater);
+                            rd.reduceMaxReadMultipleReferences();
+                            break;
+                        }
+                    }
+                    else
+                        throw new BACnetException("Completed " + counter + " requests. Excepted on: " + request, e);
+                }
+                catch (BACnetTimeoutException e) {
+                    BACnetError error = new BACnetError(ErrorClass.communication, ErrorCode.timeout);
+                    for (ObjectIdentifier oid : properties.keySet()) {
+                        for (PropertyReference ref : properties.get(oid))
+                            updater.increment(oid, ref.getPropertyIdentifier(), ref.getPropertyArrayIndex(), error);
+                    }
+                }
+                catch (BACnetException e) {
+                    throw new BACnetException("Completed " + counter + " requests. Excepted on: " + request, e);
+                }
+
+                if (updater.cancelled())
+                    break;
+            }
+        }
+        else
+            // If it doesn't support read property multiple, send them one at a time.
+            sendOneAtATime(localDevice, rd, refs, updater);
+
+        return propertyValues;
+    }
+
+    private static void sendOneAtATime(LocalDevice localDevice, RemoteDevice d, PropertyReferences refs,
+            RequestListenerUpdater updater) throws BACnetException {
+        LOG.info("Making property reference requests one at a time");
+        List<PropertyReference> refList;
+        ReadPropertyRequest request;
+        ReadPropertyAck ack;
+        Map<ObjectIdentifier, List<PropertyReference>> properties = refs.getProperties();
+        for (ObjectIdentifier oid : properties.keySet()) {
+            refList = properties.get(oid);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.